### PR TITLE
Publisher can publish articles

### PR DESCRIPTION
**Pull request by mantanquzzmo**
[PT-link](https://www.pivotaltracker.com/story/show/170674984)

- Adds Component and Tests
- Sets up routing through admin dashboard

Skärmavbild 2020-01-26 kl  15 56 55

### CODE REVIEW

File: src/modules/article.js [hunk 1]
```diff
@@ -24,20 +24,72 @@ const createArticle = async (title, body) => {
...
+const undoPublishArticle = async id => {
+  let headers = JSON.parse(localStorage.getItem("J-tockAuth-Storage"));
+  try {
+    let response = await axios({
+      url: `/admin/articles/${id}`,
+      headers: headers,
+      method: "put",
+      params: {
+          published: false
+      }
```
Comment (sundqvistjohan) [inline src/modules/article.js]: Should also set publisher_id to null(nil?) on undo as it's set to the current_user.id by the backend. Unless of course we want to know who published-and-unpublished it.

File: cypress/fixtures/publisher_articles.json [hunk 2]
```diff
@@ -0,0 +1,24 @@
+{
+  "articles": [
+    {
+    "id": 1,
+    "title": "Test1",
+    "body": "TestBody1"
+    },
```
Comment (sundqvistjohan) [inline cypress/fixtures/publisher_articles.json]: Suggest adding "published: false" to the test articles

File: cypress/integration/publisherCanPublishArticles.feature.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
+describe("Editor can", () => {
+  beforeEach(() => {
+    cy.server();
+    cy.route({
```
Comment (oliverochman) [inline cypress/integration/publisherCanPublishArticles.feature.js]: I would look into putting some of these requests that you make in all tests in the index.js file in a beforeEach block

File: cypress/integration/publisherCanPublishArticles.feature.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+      url: "http://localhost:3000/api/v1/admin/articles",
+      response: "fixture:publisher_articles.json"
+    });
+    cy.route({
+      method: "put",
+      url: "http://localhost:3000/api/v1/admin/articles/1?published=true",
+      response: "fixture:published_response.json"
+    });
+    cy.visit("/admin");
+    cy.get("#loginButton").click();
```
Comment (oliverochman) [inline cypress/integration/publisherCanPublishArticles.feature.js]: there is a way of extracting the login to the commands.js file

https://docs.cypress.io/api/cypress-api/custom-commands.html#Parent-Commands

File: src/modules/article.js [hunk 5]
```diff
@@ -5,8 +5,8 @@ const getCurrentArticle = async id => {
     const response = await axios.get(`/articles/${id}`);
     return response.data.article;
   } catch (error) {
-    if (error === 'Network Error') {
-      return { error: error.message }
+    if (error === "Network Error") {
+      return { error: error.message };
```
Comment (oliverochman) [inline src/modules/article.js]: how can you call `error.message` if error equals to "Network Error"?

File: src/modules/article.js [hunk 6]
```diff
@@ -24,20 +24,72 @@ const createArticle = async (title, body) => {
...
     return response;
   } catch (error) {
-    return error.response
+    return error.response;
   }
 };
 
-export { getCurrentArticle, getArticles, createArticle };
+const getUnpublishedArticles = async () => {
+  let headers = JSON.parse(localStorage.getItem("J-tockAuth-Storage"));
```
Comment (oliverochman) [inline src/modules/article.js:44]: the credentials are used in every request you added, can you extract this somehow so you don't need to have this one line in all of the methods?

**Review by oliverochman — CHANGES_REQUESTED** [review]

File: cypress/fixtures/publisher_articles.json [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    "title": "Test3",
+    "body": "TestBody3"
+    },
+    {
+    "id": 4,
+    "title": "Test4",
+    "body": "TestBody4"
+    }
+  ]
+}
```
Comment (tochman) [inline cypress/fixtures/publisher_articles.json]: Please format this json object. For readability....

File: cypress/fixtures/publisher_articles_after_publish.json [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    "title": "Test3",
+    "body": "TestBody3"
+    },
+    {
+    "id": 4,
+    "title": "Test4",
+    "body": "TestBody4"
+    }
+  ]
+}
```
Comment (tochman) [inline cypress/fixtures/publisher_articles_after_publish.json]: Please format this json object. For readability....

File: cypress/integration/publisherCanPublishArticles.feature.js (hunk 4 above)
Comment (tochman) [inline cypress/integration/publisherCanPublishArticles.feature.js]: Only use `kebab-case` for html tag id's, please

File: cypress/integration/publisherCanPublishArticles.feature.js [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+      url: "http://localhost:3000/api/v1/auth/**",
+      response: "fixture:login_publisher.json"
+    })
+    cy.route({
+      method: "GET",
+      url: "http://localhost:3000/api/v1/admin/articles",
+      response: "fixture:publisher_articles.json"
+    });
+    cy.route({
+      method: "put",
```
Comment (tochman) [inline cypress/integration/publisherCanPublishArticles.feature.js]: Consistency.

File: src/components/admin/AdminDashboard.jsx [hunk 10]
```diff
@@ -16,6 +17,11 @@ const AdminDashboard = props => {
         <CreateArticle />
       )
       break;
+    case props.authenticated && props.userAttrs && props.userAttrs.role === 'publisher':
```
Comment (tochman) [inline src/components/admin/AdminDashboard.jsx]: Can we simplify this? (not a blocker)

File: src/modules/article.js (hunk 1 above)
Comment (mantanquzzmo) [reply src/modules/article.js]: Fixed this in the backend

File: cypress/integration/publisherCanPublishArticles.feature.js (hunk 3 above)
Comment (mantanquzzmo) [reply cypress/integration/publisherCanPublishArticles.feature.js]: We will have a look at that after the talk

File: cypress/integration/publisherCanPublishArticles.feature.js (hunk 4 above)
Comment (mantanquzzmo) [reply cypress/integration/publisherCanPublishArticles.feature.js]: same answer as above

File: src/modules/article.js (hunk 5 above)
Comment (mantanquzzmo) [reply src/modules/article.js]: fixed

File: src/modules/article.js (hunk 6 above)
Comment (mantanquzzmo) [reply src/modules/article.js:44]: tried with axios.defaultheaders, did not work. I'll put this up on our chore list.

File: src/components/admin/AdminDashboard.jsx (hunk 10 above)
Comment (mantanquzzmo) [reply src/components/admin/AdminDashboard.jsx]: simplified.

File: cypress/integration/publisherCanPublishArticles.feature.js (hunk 9 above)
Comment (mantanquzzmo) [reply cypress/integration/publisherCanPublishArticles.feature.js]: is key. fixed.

File: cypress/integration/publisherCanPublishArticles.feature.js (hunk 4 above)
Comment (mantanquzzmo) [reply cypress/integration/publisherCanPublishArticles.feature.js]: fixed

File: cypress/fixtures/publisher_articles_after_publish.json (hunk 8 above)
Comment (mantanquzzmo) [reply cypress/fixtures/publisher_articles_after_publish.json]: formatted

File: cypress/fixtures/publisher_articles.json (hunk 7 above)
Comment (mantanquzzmo) [reply cypress/fixtures/publisher_articles.json]: formatted

**Pull request merged**